### PR TITLE
External Builder and get_external implemented

### PR DESCRIPTION
**Pull request by fckt**

### CODE REVIEW

File: src/externs.rs [hunk 1]
```diff
@@ -20,6 +20,17 @@ pub fn set_external(ext: Box<External>) {
 	});
 }
 
+pub fn get_external<T: External + Clone + 'static>() -> T {
+	EXTERNAL.with(|arg| {
+		let ref_cell: &RefCell<Box<External>> = arg;
```
Comment (pepyakin) [inline src/externs.rs]: I'm wondering, is it possible to change `|arg|` to `|ref ref_cell|`?

File: tests/simple_externs.rs [hunk 2]
```diff
@@ -5,43 +5,20 @@ extern crate pwasm_std;
...
-		12345.into()
-	}
-	fn blocknumber(&mut self) -> u64 {
-		123123u64
-	}
-	fn timestamp(&mut self) -> u64 {
-		123124u64
-	}
-}
+use pwasm_test::{ExternalBuilder};
```
Comment (pepyakin) [inline tests/simple_externs.rs]: Can we leave out those `{` `}`?

File: src/external.rs [hunk 3]
```diff
@@ -100,9 +103,88 @@ pub trait External {
...
+	fn storage_write(&mut self, key: &H256, value: &[u8; 32]) -> Result<(), Error> {
+		self.storage.insert(*key, value.clone());
+		Ok(())
+	}
+
+	fn sender(&mut self) -> Address {
+		self.sender
+	}
+
+	/// Invoked when contract is requesting coinbase extern
```
Comment (pepyakin) [inline src/external.rs]: These comments seems redudant to me, because they are repeating those in `External`.

File: src/builder.rs [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+			timestamp: 0u64,
+		}
+	}
+
+	pub fn sender(mut self, sender: Address) -> Self {
+		self.sender = sender;
+		self
+	}
+
+	pub fn storage_write(mut self, key: H256, value: [u8; 32]) -> Self {
```
Comment (NikVolf) [inline src/builder.rs]: would prefer to call it just `storage`

File: src/builder.rs [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+			address: self.address,
+			coinbase: self.coinbase,
+			difficulty: self.difficulty,
+			gas_limit: self.gas_limit,
+			blocknumber: self.blocknumber,
+			timestamp: self.timestamp,
+		}
+	}
+
+	pub fn from_instance(instance: ExternalInstance) -> ExternalBuilder {
```
Comment (NikVolf) [inline src/builder.rs]: could be `impl From<ExternalInstance> for ExternalBuilder `

File: tests/check_balance.rs [hunk 6]
```diff
@@ -5,35 +5,12 @@ extern crate pwasm_std;
...
-		self.balances[address]
-	}
-}
+use pwasm_test::ExternalBuilder;
 
 test_with_external!(
-	DummyExternal::with_balances(),
+	ExternalBuilder::new()
+		.balance_of(Address::from([1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]), 200000.into())
+	.build(),
```
Comment (NikVolf) [inline tests/check_balance.rs]: `build()` should have some indentation

File: tests/simple_externs.rs [hunk 7]
```diff
@@ -5,43 +5,20 @@ extern crate pwasm_std;
...
+		.sender("0x16a0772b17ae004e6645e0e95bf50ad69498a34e".into())
+		.address("0x35da6abcb08f2b6164fe380bb6c47bd8f2304d55".into())
+		.origin("0x51f9c432a4e59ac86282d6adab4c2eb8919160eb".into())
+		.coinbase("0xc257274276a4e539741ca11b590b9447b26a8051".into())
+		.difficulty(123.into())
+		.gas_limit(1234.into())
+		.value(12345.into())
+		.blocknumber(123123u64)
+		.timestamp(123124u64)
+	.build(),
```
Comment (NikVolf) [inline tests/simple_externs.rs]: ditto

**Comment by NikVolf** [comment]
great job, just some minor issues

**Comment by NikVolf** [comment]
also test example with inspecting external state *after* test run would be nice right here

File: src/externs.rs (hunk 1 above)
Comment (fckt) [reply src/externs.rs]: I'll check )

File: src/builder.rs (hunk 5 above)
Comment (fckt) [reply src/builder.rs]: it is there few lines below, `from_instance` is just an alias for convenience

**Comment by NikVolf** [comment]
Cargo.lock should not be here

**Comment by fckt** [comment]
@NikVolf  explain?

**Comment by fckt** [comment]
@NikVolf  ok, but maybe better to use semver.. pwasm-std interface changing from time to time..

**Comment by fckt** [comment]
> also test example with inspecting external state after test run would be nice right here
https://github.com/paritytech/pwasm-test/pull/14/commits/1b5dcf43bbe466ed681e388fbddb7b18e49f5fe2#diff-ac1fb4b8a2e77353fde52407b8606587R18

**Comment by NikVolf** [comment]
needs rebase

**Comment by NikVolf** [comment]
lgtm otherwise

**Pull request closed without merging**